drivers/platform/maxim/max32670/maxim_uart.c

```diff
@@ -0,0 +1,402 @@
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include <errno.h>
+#include <stdlib.h>
+#include "maxim_irq.h"
+#include "maxim_uart.h"
+#include "mxc_sys.h"
+#include "mxc_errors.h"
+#include "no_os_uart.h"
+#include "no_os_irq.h"
+#include "no_os_util.h"
+#include "no_os_lf256fifo.h"
+#include "uart.h"
+
+/**
+* @brief Descriptors to hold the state of nonblocking read and writes
+* on each port
+*/
+mxc_uart_req_t uart_irq_state[MXC_UART_INSTANCES];
+bool is_callback;
+
+static uint8_t c;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief UART receive interrupt callback function
+ * @param context - UART context
+ * @return none
+ */
+void uart_rx_callback(void *context)
+{
+	struct no_os_uart_desc *d = context;
+	lf256fifo_write(d->rx_fifo, c);
+	no_os_uart_read_nonblocking(d, &c, 1);
+}
+
+/**
+ * @brief Read data from UART device. Blocking function.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return positive number of received bytes in case of success, negative error code otherwise.
+ */
+int32_t no_os_uart_read(struct no_os_uart_desc *desc, uint8_t *data,
+			uint32_t bytes_number)
+{
+	int32_t ret;
+	uint32_t i = 0;
+
+	if (!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	if (desc->rx_fifo) {
+		for (i = 0; i < bytes_number; i++) {
+			ret = lf256fifo_read(desc->rx_fifo, &data[i]);
+			if (ret)
+				return i ? (int32_t)i : -EAGAIN;
+		}
+		return i;
+	}
+
+	ret = MXC_UART_Read(MXC_UART_GET_UART(desc->device_id), data,
+			    (int *)&bytes_number);
+	if (ret != E_SUCCESS)
+		return ret;
```

Comment:
This function should return the number of bytes actually read (which is a positive number). I see that unlike the FIFO implementation, this is not possible here, so returning `bytes_number` is fine by me. In case you're returning an error code, it must be a negative value defined in the `errno.h` header (which could be `-EIO` in this case) as opposed to a SDK defined err code.

Reply:
Make sense, will return -EIO from errno.h file